compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -1689,14 +1690,71 @@ static LogicalResult setRootConfig(mlir::FunctionOpInterface entryPointFn,
                                    IREE::LinalgExt::AttentionOp attnOp) {
+  IREE::LinalgExt::AttentionOpDetail opInfo(attnOp);
+  ArrayRef<int64_t> shapes = opInfo.getDimSizes();
+  TileSizesListType tileSizes;
+
+  LLVM_DEBUG({
+    KD_DBGS() << "Attention Detail:\n";
+    KD_DBGS() << "shape: [";
+    llvm::interleaveComma(opInfo.getDimSizes(), llvm::dbgs());
+    llvm::dbgs() << "]\n";
+    KD_DBGS() << "Batch: [";
+    llvm::interleaveComma(opInfo.getBatchDims(), llvm::dbgs());
+    llvm::dbgs() << "]\n";
+    KD_DBGS() << "M: [";
+    llvm::interleaveComma(opInfo.getMDims(), llvm::dbgs());
+    llvm::dbgs() << "]\n";
+    KD_DBGS() << "K1: [";
+    llvm::interleaveComma(opInfo.getK1Dims(), llvm::dbgs());
+    llvm::dbgs() << "]\n";
+    KD_DBGS() << "K2: [";
+    llvm::interleaveComma(opInfo.getK2Dims(), llvm::dbgs());
+    llvm::dbgs() << "]\n";
+    KD_DBGS() << "N: [";
+    llvm::interleaveComma(opInfo.getNDims(), llvm::dbgs());
+    llvm::dbgs() << "]\n";
+  });
+
+  // TODO (Groverkss): Flash Attention 2 (current algorithm we use for
+  // attention) was originally designed for GPUs. N, K1 are the head dimension
+  // and are usually very small (64, 128, See AttentionOpDetail docs for more
+  // detail). For larger sizes, fusing attention doesn't have many gains (as
+  // pointed out by the original author). We should explore if we should tile N
+  // and K1 dimensions on CPU and if it has any gains. On GPUs, we don't tile
+  // these dimensions as subgroups can hold much larger register sizes.
+
+  // Batch, M and N (parallel dimensions) are distributed on workgroups.
+  DistributionHeuristicConfig config;
   SmallVector<int64_t> distTileSizes = getDefaultDistributedLevelTileSizes(
       attnOp, DistributionHeuristicConfig{});
-  int64_t iterationDomainRank = attnOp.getIterationDomainRank();
-  SmallVector<int64_t> vecTileSizes(iterationDomainRank, 1);
+  tileSizes.push_back(distTileSizes);
+
+  // Batch, M and N (parallel dimensions) are distributed on workgroups.
+  SmallVector<int64_t> vecTileSizes(opInfo.getDomainRank(), 1);
+  // Mark reduction dimensions not to distribute.
+  for (int64_t i :
+       llvm::concat<const int64_t>(opInfo.getK1Dims(), opInfo.getK2Dims())) {
+    vecTileSizes[i] = 0;
+  }
   int64_t vectorSize = getVectorSize(entryPointFn, attnOp.getOutputType());
-  vecTileSizes.back() = getMaxVectorTileSize(
-      /*numElem=*/distTileSizes.back(), vectorSize, vectorSize);
-  TileSizesListType tileSizes = {distTileSizes, vecTileSizes};
+  for (auto i : llvm::seq<unsigned>(0, vecTileSizes.size())) {
+    // Do not tile reduction dimensions.
+    if (vecTileSizes[i] == 0) {
+      continue;
+    }
+    auto tileSize = distTileSizes[i] ? distTileSizes[i] : shapes[i];
+    // TODO: Ask Hanhan what the native size problem here is and change the
+    // "tileSize" arguement to getMaxVectorTileSize accordingly.
+    vecTileSizes[i] = getMaxVectorTileSize(
+        /*numElem=*/tileSize, vectorSize, vectorSize);
```

Comment:
native_vector_size usually means the number of bits of a register on the target cpu. We need to change the `tileSize` to be a factor of problem size because we want to vectorize it without masking. It is a workaround to bypass scf.for bufferization issue.

Reply:
Ah ok makes sense.